trieste/models/keras/utils.py

```diff
@@ -98,3 +102,62 @@ def negative_log_likelihood(
     :return: Negative log likelihood values.
     """
     return -y_pred.log_prob(y_true)
+
+
+class KernelDensityEstimator:
+    """
+    Kernel density estimator of the D-dimensional parameter space.
+    """
+    def __init__(self, kernel: str = 'gaussian'):
+        self.kernel = kernel
+        
+    def fit(self, query_points: tf.Tensor, bandwidth: int = None):
+        """
+        Fits an optimal bandwidth given a number of query points. We use a grid search to find
+        the optimal value in the bandwidth space, defined as the bandwidth that maximizes the
+        in-sample posterior probability of observing the points. 
+
+        Given a bandwidth, individual Gaussian distributions are constructed for each observation.
+
+        :param query_points: The points to use while searching the optimal bandwidth.
+        :param bandwidth: A fixed bandwidth can be instead provides, by default is None.
+        """
+        if bandwidth is not None:
+            self.bandwidth = bandwidth
+        else:
+            if query_points.shape[0] > 1:
+                bandwidth_search_space = np.logspace(-3,1,50)
+                grid = GridSearchCV(
```

Comment:
And as per your other comment above, I've changed it now so that we can also benchmark on CV loops of init_buffer (CV is set to 2 tho, otherwise computational cost just increases too much)